[rocthrust] Edit the reference section of the documentation. (#6145)
ROCm/rocm-libraries: projects/rocthrust/docs/reference/rocThrust-bitwise-repro.rst

rocThrust bitwise reproducibility
************************************

> When the HIP backend is used, the default ``thrust::device`` policy dispatches to the ``thrust::hip::par`` policy.

Reviewer: I'm not sure about the accuracy of policies "dispatching to" other policies. I'd suggest maybe something like "the default ``thrust::device`` policy is set up as an alias for the ``thrust::hip::par`` policy."

> Not all rocThrust functions are bitwise reproducible under the ``thrust::hip::par`` policy. The following functions are bitwise reproducible for associative operators but not for for pseudo-associative floating point operators: 

Reviewer: There's a double "for" in the last sentence in this paragraph.

* ``inclusive_scan``
* ``exclusive_scan``
* ``inclusive_scan_by_key``
* ``exclusive_scan_by_key``
* ``transform_inclusive_scan``
* ``transform_exclusive_scan``
* ``reduce_by_key``

Bitwise reproducible versions of these functions for pseudo-associative floating point operators are available under the ``thrust::hip::par_det`` deterministic parallel execution policy. 

> When using these functions with ``thrust::hip::par_det``, their bitwise reproducibility must be verified regardless of the type of operator. Because of this overhead, these functions should only use ``thrust::hip::par`` with associative operators. 

Reviewer: I don't think "must be verified" is accurate in the first sentence here - it's more like it must be enforced. Here's a suggestion for how it could be reworded with some more detail:

"When using these functions with ``thrust::hip::par_det``, operations are forced to complete in a fixed order. This ensures that the result is bitwise-identical on every run."

.. note::

>     The behavior of other bitwise reproducible functions under the ``thrust::hip::par_det`` policy will be identical to their behavior under the default policy, and they can be used under either policy without the need for testing.

Reviewer: I would suggest removing the second half of the sentence here ("and they can be used under either policy without the need for testing" because it gives the impression that results will be consistent from run to run. The behaviour under the default policy is to allow blocks to complete in different orders on different runs. This may or may not affect bitwise reproducibility - it depends on what data type they're using.

To run the bitwise reproducibility tests, first build the ``reproducibility.hip`` target:

.. code:: shell

       cmake --build build --target reproducibility.hip

.. note::

    rocThrust must have been built with ``-DBUILD_TEST=ON`` to build ``reproducibility.hip``.

This target tests bitwise reproducibility either by issuing multiple calls to the functions or by running multiple iterations of the same test.

In the first case, where multiple calls are made, a special scan operator inserts a random amount of delay into calculations to create variations in the internal timing of operations. The test then verifies that the results for each call are the same. All calls are issued within a single run of the test program.

> In the second case, several runs of the test are done and compared to each other. On the first run, the test stores all the input and output pairs for each function in a database file. On subsequent runs, the test compares the input and output pairs to those in the database. If identical pairs for a function are found, the test has succeeded. If no matching pair is found, the test has failed.

Reviewer: Technically, it would be a little more accurate to say that the test stores *hashes of* the inputs and outputs, not the inputs and outputs themselves (which could be very large and require a lot of space), but this probably isn't a huge deal here. Feel free to ignore this comment if you like.

Author: I'm going to change the language a bit. It won't say "hash" but it'll be a bit less definitive.

> To enable this test, set ``ROCTHRUST_BWR_PATH`` to the path to the database file. Set ``ROCTHRUST_BWR_GENERATE`` to 1 on the first run of the test, and to 0 for all subsequent runs.

Reviewer: It might be helpful to make clearer that setting ``ROCTHRUST_BWR_GENERATE=1`` will create the database file, though I guess the name communicates this already.

Author: Going to make it clearer.

.. code:: shell

    ROCTHRUST_BWR_PATH=/path/to/repro.db ROCTHRUST_BWR_GENERATE=1 reproducibility.hip
    ROCTHRUST_BWR_PATH=/path/to/repro.db reproducibility.hip

> A new first iteration of the test must be run if the ROCm version, rocThrust version, or GPU architecture changes.

Reviewer: It might be helpful to expand on this a bit - if you change any of those 3 things, then you need to generate a new database file by setting ``ROCTHRUST_BWR_GENERATE=1`` (and changing ``ROCTHRUST_BWR_PATH`` if you don't want your old database file to get overwritten).

* ``exclusive_scan_by_key``
* ``transform_inclusive_scan``
* ``transform_exclusive_scan``
* ``reduce_by_key``

.. note:: 

    Under the :doc:`HIP backend <../how-to/rocThrust-build-for-backends>`, the default ``thrust::device`` policy is an alias for the ``thrust::hip::par`` policy.

Bitwise reproducible versions of these functions for pseudo-associative floating point operators are available under the ``thrust::hip::par_det`` deterministic parallel execution policy. 

When using these functions with ``thrust::hip::par_det``, operations are forced to complete in a fixed order. This ensures that the result is bitwise-identical on every run. Because of this overhead, the ``thrust::hip::par_det`` policy should only be used with pseudo-associative floating point operators.  The ``thrust::hip::par`` policy should be used otherwise.

.. note::

    The behavior of other bitwise reproducible functions under the ``thrust::hip::par_det`` policy will be identical to their behavior under the ``thrust::hip::par`` policy.

To run bitwise reproducibility tests, first build the ``reproducibility.hip`` target:

.. code:: shell

       cmake --build build --target reproducibility.hip

.. note::

>     rocThrust must have been built with ``-DBUILD_TEST=ON`` to build ``reproducibility.hip``.

Reviewer 2: Is this right?
Must have been built?

Author: Yes. In the past, you needed to have built it this way. Otherwise this won't build.

This target tests bitwise reproducibility either by issuing multiple calls to the functions or by running multiple iterations of the same test.

In the first case, where multiple calls are made, a special scan operator inserts a random amount of delay into calculations to create variations in the internal timing of operations. The test then verifies that the results for each call are the same. All calls are issued within a single run of the test program.

> In the second case, several runs of the test are done and compared to each other. On the first run, the test stores input and output pairs for each function in a database file. On subsequent runs, the test compares the input and output pairs to those in the database. If identical pairs for a function are found, the test has succeeded. If no matching pair is found, the test has failed.

Reviewer 2: ```suggestion
In the second case, several test runs are performed and compared. On the first run, the test stores input-output pairs for each function in a database file. In subsequent runs, the test compares the input-output pairs to those in the database. If identical pairs for a function are found, the test has succeeded. If no matching pair is found, the test has failed.
```
